### ISSUE TITLE
Custom repository: "Undefined index: uid"

### ISSUE THREAD

**Issue by lukasbestle** [composer/composer#7366]
I'm currently building a small custom Composer repository for private packages and need a bit of help with the most optimal setup:

1. I'm assuming that using a `provider` file for every package instead of cramming all data into the main `packages.json` file increases performance as packages that haven't changed can be cached based on their SHA256 hash. Is this correct? Or doesn't it matter?
2. The [docs](https://getcomposer.org/doc/05-repositories.md#composer) state that a package definition in the repo only needs the `name`, the `version` and a `dist` or `source`. Is it useful to include any further information (i.e. are any other fields used by the client and if so, how)? Which fields are recommended?
3. I was getting the error `Undefined index: uid` as I didn't include a UID in the package definition in my repo. This field is not documented in the [schema](https://getcomposer.org/doc/04-schema.md), but because of #2738 I was able to figure out that this field should be unique per repository. Is it alright to use a string of a format like `"<vendor>/<package>@<version>"` or might this cause issues? As I don't know how the UID is used internally I can't really tell which characteristics that field must have.

Thank you in advance! :)

**Comment by alcohol** [comment]
Not sure your question belongs here. While you can inline custom repository definitions in your composer.json, those are not quite the same as for example running a [satis](https://github.com/composer/satis) instance or a private instance of [packagist](https://github.com/composer/packagist). The schema you mentioned also does not apply to those, but rather Composer, and specifically, the composer.json file.

**Comment by lukasbestle** [comment]
Thanks for your replies! :)
Maybe there was a misunderstanding, so I will try to make a bit clearer what I'm trying to do:

> While you can inline custom repository definitions in your composer.json

That's not what I'm trying to do. I want to host a Composer repository on a web server just like with Satis.

The only difference is that I won't be able to use Satis for this and therefore want to know what I need to take care of when generating the repository JSON files myself. I got my custom repo working by referring to the [repo docs](https://getcomposer.org/doc/05-repositories.md#composer), but still have the four questions which are not answered by the docs.

> the schema for files in a composer repository (like Satis or Packagist) is not the same than this schema. It includes more info.

According to the [repo docs](https://getcomposer.org/doc/05-repositories.md#composer), a package definition in a repo "may include any of the other fields specified in the [schema](https://getcomposer.org/doc/04-schema.md)". There is no place in the docs on getcomposer.org where the additional fields of the repository schema are documented, which is why I asked my four questions.

---

If this repo is not the right place to ask my questions, where can I ask them?
If this repo *is* the right place, I would be very grateful to get those specific questions answered. Thanks in advance!

**Comment by lukasbestle** [comment]
I already dug through the Satis code, but unfortunately the implementation doesn't really tell me *why* the schema is like it is, so I can't really get any information that answers my questions.

I know my questions are quite specific, but it would help me a lot to get a direct answer to each of them.

**Comment by stof** [comment]
> Is this correct? Or doesn't it matter?

yes, allowing separate caching is the goal of this provider thing. 
The other advantage is that your project may not need all packages, and so can skip loading of some files entirely (this is especially relevant for Packagist, where you probably won't ever find a single project using the 183 481 packages which registered right now)

> Are outdated provider files (i.e. those with the "old" hash once a new provider file with a new hash gets published) ever accessed again?

they might get accessed in case of race condition between an update of your repository and client-side usage (loading the old packages.json and then loading some of the provider files after the server dumped the update). What Packagist does is that it keeps the previous provider files around when performing an update, and cleans them on the next update (as it runs an update every 5 minutes, it deletes all files which were not touched during the last 10 minutes)

> Is it useful to include any further information (i.e. are any other fields used by the client and if so, how)? Which fields are recommended?

Most package metadata are needed client-side (otherwise, composer would not have them at all). IIRC, packagist currently drops the `support` key (and maybe a few others) which are only there to provide some human-readable info (and this is reported as an issue as it means `composer show` won't display them). Here is the code building the representation: https://github.com/composer/packagist/blob/d162ebbb929aacccf8a5ea4fe9a646189130f154/src/Packagist/WebBundle/Entity/Version.php#L214-L292 
 https://github.com/composer/packagist/blob/d162ebbb929aacccf8a5ea4fe9a646189130f154/src/Packagist/WebBundle/Package/SymlinkDumper.php#L623-L625

> As I don't know how the UID is used internally I can't really tell which characteristics that field must have.

The UID is used in composer in case the same package exists in multiple provider files (this happens in case you use `provide` or `replace` as the package has to appear in the provider files for these names too), to allow composer to know that they are the same package and not 2 different packages. `<vendor>/<package>@<version>` might work for that (be sure to now have any `@` in your branch names, otherwise the dev version for such branches might create ambiguities in this pattern). Packagist uses its database ID instead, allowing the JSON to be more compact.

**Comment by alcohol** [comment]
Also, https://github.com/composer/composer/blob/master/res/composer-repository-schema.json does specify what the schema should look like, to some extend. I am not sure how up-to-date that is though or if we ever made it "complete".
